src/libvirtApi/domain.js

```diff
@@ -1112,3 +1113,23 @@ export function domainUpdateDiskAttributes({ connectionName, objPath, target, re
                 return call(connectionName, '/org/libvirt/QEMU', 'org.libvirt.Connect', 'DomainDefineXML', [updatedXML], { timeout, type: 's' });
             });
 }
+
+export async function domainReplaceSpice({ connectionName, id: objPath }) {
+    /* Ideally this would be done by virt-xml, but it doesn't offer that functionality yet
+     * see https://issues.redhat.com/browse/RHEL-17436 */
+    const [domXML] = await call(connectionName, objPath, 'org.libvirt.Domain', 'GetXMLDesc', [Enum.VIR_DOMAIN_XML_INACTIVE], { timeout, type: 'u' });
+    const updatedXML = replaceSpice(domXML);
+
+    // check that updatedXML is valid; if not, throw; it needs to be updated manually
+    await (cockpit.spawn(["virt-xml-validate", "-"], { err: "message" }).input(updatedXML));
+
+    await call(connectionName, objPath, 'org.libvirt.Domain', 'Undefine', [0], { timeout, type: 'u' });
```

Comment:
Is there no way to update the XML of a existing domain?

Reply:
Nope. `virsh edit` may do this in a more robust way (doing undefine/redefine much closer together), but that's not accessible through the API.

Reply:
@martinpitt `Define` works for an existing VM, virt-xml doesn't do anything with `undefine` for example. Were you hitting a specific error?

Reply:
@crobinso Indeed, nice! I don't remember any more where/how I saw this, but indeed both `virsh define` as well as the API seem to do this right -- if it fails due to invalid XML, the previous definition stays intact. That makes this much more robust. I dropped this, and simplified the cleanup below.

Reply:
Yay!